File: Assignment_1/task1.py
from pyspark import SparkContext
import sys
import time
import json

start_time = time.time()

input_file = ''
output_file = ''
stop_words = ''
y = 0
m = 0
n = 0

special_chars_set = {'.',',','!','?',';',':',')',']','(','['}
stop_words_set={}

class Helper:

	def part_D_tuple_to_list(users):
		users_list = []
		for i in users:
			users_list.append(list(i))
		return users_list

	def part_E_to_list(text):
		review_list = []
		for i in text:
			review_list.append(i[0])
		return review_list

	def reformat(adul_word):
		word_req = ""
		for character in adul_word:
			if character not in special_chars_set:
				word_req += character
		word_modified = ''.join(word_req)
		if word_modified not in stop_words_set:
			return word_modified


class Review:

	def part_A(self, rdd_lines):
		row_count = rdd_lines.count()

		return row_count

	def part_B(self, rdd_lines, y):
		rdd_columns = rdd_lines.map(lambda data_row: (data_row['date'], data_row['review_id']))
		rdd_matched_years = rdd_columns.filter(lambda data_row: int(data_row[0][:4])==y)

		return rdd_matched_years.count()

	def part_C(self, rdd_lines):
		rdd_columns = rdd_lines.map(lambda data_row: data_row['business_id'])
		rdd_matched_years = rdd_columns.distinct()

		return rdd_matched_years.count()

	def part_D(self, rdd_lines, m):
		rdd_columns = rdd_lines.map(lambda data_row: (data_row['user_id'], data_row['review_id']))
		rdd_user_pair = rdd_columns.map(lambda data_row: (data_row[0], 1))
		rdd_reduce = rdd_user_pair.reduceByKey(lambda key1,key2: key1+key2).sortByKey(ascending=True)
		rdd_reduce_top_m = rdd_reduce.takeOrdered(m, key=lambda count_reviews: (-count_reviews[1], count_reviews[0]))
		required_list = Helper.part_D_tuple_to_list(rdd_reduce_top_m)

		return required_list

	def part_E(self, rdd_lines,n):
		rdd_columns = rdd_lines.map(lambda data_row: data_row['text'])
		rdd_one_list = rdd_columns.flatMap(lambda data_row: data_row.lower().split())
		rdd_text_pair = rdd_one_list.map(lambda data_row: (Helper.reformat(data_row),1))
		rdd_pair_filter = rdd_text_pair.filter(lambda key:(key[0] != None and key[0] != ""))
		rdd_reduce = rdd_pair_filter.reduceByKey(lambda key1, key2: key1+key2).sortByKey(ascending=True)
		rdd_reduce_top_n = rdd_reduce.takeOrdered(int(n), key=lambda freq:(-freq[1], freq[0]))
		required_list = Helper.part_E_to_list(rdd_reduce_top_n)

		return required_list

class Building_Structure():

	def read_input(self):
		global input_file, output_file, stop_words, y, m, n
		input_file = sys.argv[1]
		output_file = sys.argv[2]
		stop_words = sys.argv[3]
		y = sys.argv[4]
		m = sys.argv[5]
		n = sys.argv[6]

	def process_output(self, part_A, part_B, part_C, part_D, part_E):
		output_dict = {}
		output_dict['A'] = part_A
		output_dict['B'] = part_B
		output_dict['C'] = part_C
		output_dict['D'] = part_D
		output_dict['E'] = part_E

		# print(output_dict)
		f = open(output_file,"w")
		json.dump(output_dict,f)
		f.close()

def main():
	global stop_words_set 
	building_structure = Building_Structure()

	building_structure.read_input()
	# print("Arguments Passed: ", input_file, output_file, stop_words, y, m, n)

	#creating Object
	sc_object = SparkContext()
	rdd_lines = sc_object.textFile(input_file).map(lambda line: json.loads(line))

	calculate_results = Review()

	part_A_ans = calculate_results.part_A(rdd_lines)
	part_B_ans = calculate_results.part_B(rdd_lines, int(y))
	part_C_ans = calculate_results.part_C(rdd_lines)
	part_D_ans = calculate_results.part_D(rdd_lines, int(m))
	file = open(stop_words,"r")
	stop_words_set = set(word.strip() for word in file)
	part_E_ans = calculate_results.part_E(rdd_lines, int(n))

	building_structure.process_output(part_A_ans, part_B_ans, part_C_ans, part_D_ans, part_E_ans)


if __name__ == "__main__":
	main()

# print("--- %s seconds ---" % (time.time() - start_time))

File: Assignment_1/task1.scala
import java.io.{PrintWriter, File}

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import org.json4s._
import org.json4s.jackson.JsonMethods._

import scala.io.Source._

object task1 {

  case class Review_Class(review_id: String, user_id: String, business_id: String, stars: String, text: String, date: String)

  def main(args: Array[String]) {

    val sparkSession = SparkSession.builder().appName("task1").config("spark.master", "local[*]").getOrCreate()

    val context = sparkSession.sparkContext
    context.setLogLevel("ERROR")
    val inputFile = args(0)
    val outputFile = args(1)
    val stopWords = args(2)
    val y = args(3)
    val m = args(4)
    val n = args(5)

    val lines = fromFile(stopWords).getLines.toArray
    val list_stop = lines
    val strip_l = list_stop.map(_.trim)

    val rdd = context.textFile(inputFile).
      map(extractData).map(x => (x.business_id, x.date, x.review_id, x.stars, x.text, x.user_id))

    val part_A_ans = part_A(rdd)
    val part_B_ans = part_B(rdd, y)
    val part_C_ans = part_C(rdd)
    val part_D_ans = part_D(rdd, m)
    val part_E_ans = part_E(rdd, n, strip_l)
    process_output(part_A_ans, part_B_ans, part_C_ans, part_D_ans, part_E_ans, outputFile)
  }

  def process_output(partA: Long, partB: Long, partC: Long, partD: Array[(String, Int)], partE: Array[(String, Int)], output: String) = {
    println(partA, partB, partC, partD.mkString(""), partE.mkString(""))
    val writerObject = new PrintWriter(new File(output))
    writerObject.write('{')
    writerObject.write("\"A\": " + partA + ", ")
    writerObject.write("\"B\": " + partB + ", ")
    writerObject.write("\"C\": " + partC + ", ")
    writerObject.write("\"D\": [")
    for (i <- partD) {
      writerObject.write("[\"" + i._1 + "\", " + i._2 + "]")
      if (partD.indexOf(i) < partD.length - 1) {
        writerObject.write(",")
      }
    }
    writerObject.write("],")
    writerObject.write("\"E\": [")
    for (i <- partE) {
      writerObject.write("\"" + i._1 + "\"")
      if (partE.indexOf(i) < partE.length - 1) {
        writerObject.write(", ")
      }
    }
    writerObject.write("]}")
    writerObject.close()

  }

  def reformat(word: String, stop: Array[String]): String = {

//    val rem = word.replaceAll("[\\t\\n]", " ").trim()

    val remove_sp_ch = word.replaceAll("[.,!?;:)\\](\\[)]", "").trim()


    if (stop.contains(remove_sp_ch)) {
      return ""
    }
    else {
      return remove_sp_ch
    }
  }

  def part_A(rdd: RDD[(String, String, String, String, String, String)]): Long = {
    val rdd1 = rdd.map(x => (x._3))
    return rdd1.count()
  }

  def part_B(rdd: RDD[(String, String, String, String, String, String)], y: String): Long = {
    val rdd1 = rdd.map(x => (x._3, x._2))
    val rdd2 = rdd1.map(x => x._2)
    val rdd_filter_year = rdd2.filter(x => x.contains(y))
    return rdd_filter_year.count()
  }

  def part_C(rdd: RDD[(String, String, String, String, String, String)]): Long = {
    val rdd1 = rdd.map(x => x._1).distinct()
    return rdd1.count()
  }

  def part_D(rdd: RDD[(String, String, String, String, String, String)], m: String): Array[(String, Int)] = {
    val rdd1 = rdd.map(x => (x._6, x._3))
    val rdd_pair_kv = rdd1.map(x => (x._1, 1))
    val rdd_reduce = rdd_pair_kv.reduceByKey(_ + _).sortByKey(ascending = true)
    val top_m = rdd_reduce.takeOrdered(Integer.parseInt(m))(Ordering.by(x => (-x._2, x._1)))
    return top_m
  }

  def part_E(rdd: RDD[(String, String, String, String, String, String)], n: String, stop: Array[String]): Array[(String, Int)] = {
    val rdd1 = rdd.map(x => (x._5))

    val rdd_flat = rdd1.flatMap(x => x.toLowerCase.split(" "))
    val rdd_pair = rdd_flat.map(x => (reformat(x, stop), 1))
    val rdd_filter = rdd_pair.filter(x=>x._1.length!=0)
    val rdd_reduce = rdd_filter.reduceByKey(_ + _).sortByKey(ascending = true)
    val top_m = rdd_reduce.takeOrdered(Integer.parseInt(n))(Ordering.by(x => (-x._2, x._1)))
    return top_m
  }

  def extractData(line: Any): Review_Class = {
    implicit val formats: DefaultFormats.type = DefaultFormats
    parse(line+"").extract[Review_Class]
  }
}
